examples/arc_01/run_pstd_bscan.m

```diff
@@ -1,64 +1,151 @@
-%setup free space matrix and save to directory gridfiles/
-composition_matrix = [];
-save('gridfile_fs', 'composition_matrix', 'material_matrix');
+% Now produce the material file for the freespace simulation
+composition_matrix = composition_matrix_builder(input_filename, obstacle_fs, 0.);
+material_matrix = [1 refind^2 1 0 0 0     0     0     0 0 0];
+gridfile_fs = 'gridfile_fs.mat';
+save(gridfile_fs, 'composition_matrix', 'material_matrix', '-v7.3');
+
+
+%% 3.2.2 Source file specification
+% If our input_filename did not specify that we wish to use a matlab
+% function to compute the incident fields on the interface planes, we could
+% provide a pre-made .mat file containing this information to
+% iteratefdtd_matrix by populating the illfile variable.
+% Since we do not wish to do this, we leave the name empty and
+% iteratefdtd_matrix will infer from the input file that we want to compute
+% the source fields from scratch.
+illfile = '';
 
-%generate tdms executable input files
-[fdtdgrid, Exs,Eys,Ezs,Hxs,Hys,Hzs,grid_labels,camplitudes,vertices,facets,Id] = iteratefdtd_matrix('pstd_input_file.m','filesetup','in_pstd_cyl','gridfile_cyl.mat','');
-[fdtdgrid, Exs,Eys,Ezs,Hxs,Hys,Hzs,grid_labels,camplitudes,vertices,facets,Id] = iteratefdtd_matrix('pstd_input_file.m','filesetup','in_pstd_fs','gridfile_fs.mat','');
+%% Use iteratefdtd_matrix to setup the input file(s).
+% Create the input to the simulation that models scattering from a
+% cylindrical obstacle.
+% Write the created input file to the cylinder_input.mat file.
+iteratefdtd_matrix(input_filename,'filesetup','cylinder_input.mat',gridfile_cyl,illfile);
+% Create the input to the simulation that has no obstacle, and only models
+% free space.
+% Write the created input file to the freespace_input.mat file.
+iteratefdtd_matrix(input_filename,'filesetup','freespace_input.mat',gridfile_fs,illfile);
 
-%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
-%Now run the executeables
-%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Run the tdms executable
+% These commands can be carried out via the CLI
+% Because MATLAB ships with versions of libstdc++ and prepends these to the
+% path when calling system, when running tdms from MATLAB it is necessary
+% to overwrite the search path for the libraries we need.
+% TDMS is built in such a way that it does not rely on environment
+% variables being present to run.
+% ON UNIX: You require the 'LD_LIBRARY_PATH', '' option
+% ON MAC: You require the 'DYLD_LIBRARY_PATH', '' option
+% ON WINDOWS: You should not require any additional arguments
 
-eval('!tdms in_pstd_fs.mat out_pstd_fs.mat');
-eval('!tdms in_pstd_cyl.mat out_pstd_cyl.mat');
+% Run the simulation for the freespace setup
+% On the command line, in the examples/arc_01 folder, run
+% tdms freespace_input.mat freespace_output.mat
+system('tdms freespace_input.mat freespace_output.mat', 'LD_LIBRARY_PATH', '', 'DYLD_LIBRARY_PATH', '');
```

Comment:
@samcunliffe Can you check that running this on MacOS does what you expect?

If it does/does not, can you also check if just running
```matlab
system('tdms freespace_input.mat freespace_output.mat', 'LD_LIBRARY_PATH', '');
```
on this line works too? We might only need the `LD_LIBRARY_PATH` thing on Linux.

Reply:
Yeah so MacOS/MATLAB R2022a gives me
```
>> run_pstd_bscan
Allocating grid...Done
Initialising source field...
Creating Isource, Jsource, Ksource...Done initialising source field
Initialising update terms [free space and pml]...Done
Initialising update terms [reading grid composition from gridfile_cyl.mat]...Done
Writing cylinder_input.mat...Done (iteratefdtd_matrix)
Allocating grid...Done
Initialising source field...
Creating Isource, Jsource, Ksource...Done initialising source field
Initialising update terms [free space and pml]...Done
Initialising update terms [reading grid composition from gridfile_fs.mat]...

Warning: An non-empty material matrix has been specified 
along with an empty composition matrix - the material
matrix will not be integrated into the FDTD input file.

Done
Writing freespace_input.mat...Done (iteratefdtd_matrix)
Error using run_pstd_bscan
Too many input arguments.
```

If I remove the `'DYLD_LIBRARY_PATH'` I get the same processing and it complains with:
```
Error using run_pstd_bscan
Unrecognized option: LD_LIBRARY_PATH
 ```

Reply:
> If I remove the `'DYLD_LIBRARY_PATH'` stuff,

I think you removed too much, the syntax is
```
system(executable, variable_name1, variable_value1, variable_name2, variable_value2, etc...)
```
so you're missing the third argument.  Or just remove `'LD_LIBRARY_PATH'` altogether for your test.

Reply:
Okay, and
```MATLAB

% Run the simulation for the freespace setup
% On the command line, in the examples/arc_01 folder, run
% tdms freespace_input.mat freespace_output.mat
system('tdms freespace_input.mat freespace_output.mat');
% Run the simulation for the cylindrical scatterer
% On the command line, in the examples/arc_01 folder, run
% tdms cylinder_input.mat cylinder_output.mat
system('tdms cylinder_input.mat cylinder_output.mat');
```
Doesn't work with a 
```
zsh:1: command not found: tdms
zsh:1: command not found: tdms
```

However in my terminal I _do_ have TDMS installed...
```sh
➤ tdms
Incorrect number of arguments. See below for help

Usage:
tdms [options] infile outfile
tdms [options] infile gridfile outfile
Options:
-h:	Display this help message
-v, --version:	Display tdms version information
-q:	Quiet operation. Silence all logging
-m:	Minimise output file size by not saving vertex and facet information

➤ echo $SHELL
/bin/zsh
➤ which tdms
/usr/local/bin/tdms
```

And indeed if I force the full path...
```MATLAB
system('/usr/local/bin/tdms freespace_input.mat freespace_output.mat');
```
It runs and I see the lovely plots.

Reply:
> Warning: An non-empty material matrix has been specified 
> along with an empty composition matrix - the material
> matrix will not be integrated into the FDTD input file.

This is true of all the input generation... it's essentially saying that "without an obstacle, I can't make sense of your definition of a scattering object, so I'll assume you want a homogeneous medium (no scatterer)".

This is build into `iteratefdtd_matrix`, specifically in the `read_material_data` function.

Reply:
> You're missing the extra `''` argument at the end of the function call

Sorry (slow to get notifications). Even with the extra empty string...
```MATLAB
system('tdms freespace_input.mat freespace_output.mat', 'LD_LIBRARY_PATH', '');
```
I get 
```
Error using run_pstd_bscan
Unrecognized option: LD_LIBRARY_PATH
```

Reply:
> 

Great... you can run
```matlab
system('which(tdms)')
```
in the MATLAB prompt to see what MATLAB's session believes the path to tdms is. You can also
```matlab
system('echo $PATH')
```
to see what it's done to your machine.

Reply:
Yeah it's removed `/usr/local/bin` from my `PATH`.

Reply:
At this point, is it going to be the path of least resistance to have 3 example scripts just for these two lines? :sob: (assuming we can get this thing working ofc)

Reply:
> How are you starting Matlab? From the launcher or from the terminal?

Great minds think alike... 

Terminal > `/Applications/MATLAB_R2022a.app/bin/matlab` and I now have `tdms`.
So I can run
```MATLAB
system('tdms freespace_input.mat freespace_output.mat');
```
and see the expected iterating debug and plots.

Still can't get it to work with the `'LD_LIBRARY_PATH', ''` arguments though.

Reply:
So:

* Starting MATLAB from the MacOS launcher doesn't preserve the `PATH` variable. Starting from a terminal does.
* MacOS doesn't accept extra arguments to `system` but Ubuntu needs that to work?

Does it preserve the `LD_LIBRARY_PATH` if you start MATLAB from the terminal in Ubuntu?

Reply:
> Still can't get it to work with the 'LD_LIBRARY_PATH', '' arguments though.

Which is odd since on Ubuntu it runs completely fine even though `DYLD_LIBRARY_PATH` is not defined.

So the path of least resistance: we have 2 demo scripts since Ubuntu clearly gets special treatment by MATLAB?

Still doesn't answer how we want to handle the problem that, even if `tdms` is in the user's path, the example script might not pick it up. 

I did have one idea: turn the example script into a bash/Python script instead. We know Python can reliably call tdms, and it can reliably call matlab too. We can then save the plots rather than having the matlab GUI open them at the end?

Reply:
Perhaps leave the `LD_LIBRARY_PATH` and `DYLD_LIBRARY_PATH` lines **_in_** there but commented out with instructions for what to do if running on Ubuntu. Or visa versa for MacOS.

Reply:
Something like

```python

# Call MATLAB to setup the input file

# Use subproccess to run the tdms commands

# Call MATLAB to produce the output plots

# Print happy message :rainbow: and exit
```

Reply:
> Perhaps leave the LD_LIBRARY_PATH and DYLD_LIBRARY_PATH lines in there but commented out with instructions for what to do if running on Ubuntu. Or visa versa for MacOS.


https://stackoverflow.com/questions/24923384/how-to-get-matlab-to-determine-if-the-os-is-windows-or-mac-so-to-find-all-seri

We _can_ get around this issue at least

Reply:
Nice*, so we can fix the example to work on 2/3 OSs. I'm guessing on Windows it'll be a
```MATLAB
system('tdms.exe foo bar')
```

*) Not very nice. But at least it works.

Reply:
Can we check this on a Windows Machine if we have one to hand?